File: packages/crypto/src/ijson.ts
/**
 * RFC 7493 (I-JSON) raw-bytes gate.
 *
 * Runs on the UTF-8 bytes of a JSON document BEFORE JSON.parse, because platform
 * JSON parsers can collapse duplicate object member names, round or overflow
 * numeric values, or substitute invalid string data before higher-level
 * validation sees the original bytes. The gate also classifies string-level
 * syntax failures consistently at the raw boundary. It rejects, per I-JSON
 * (RFC 7493):
 *
 *   - duplicate object member names  -> collapsed by JSON.parse (last wins)
 *     => CRYPTO_IJSON_DUPLICATE_MEMBER_NAME
 *   - numbers that are non-finite OR whose absolute magnitude exceeds 2^53-1
 *     (incl. large finite floats) -> rounded/overflowed by JSON.parse
 *     => CRYPTO_IJSON_NUMBER_OUT_OF_RANGE
 *   - invalid string content: lone surrogates or noncharacters (directly encoded
 *     or in \u escapes), invalid escape sequences, or invalid UTF-8 bytes
 *     inside a string => CRYPTO_IJSON_INVALID_STRING
 *
 * Escaped solidus ("\/") is valid I-JSON and MUST be accepted. String-encoded
 * large integers (inside quotes) are accepted. Valid surrogate PAIRS are
 * accepted. The numeric bound is exactly 9007199254740991 (= MAX_SAFE_INTEGER).
 *
 * This is a strict single-pass JSON parser operating on the RAW BYTES (it does
 * NOT pre-decode the whole document, so it can distinguish invalid UTF-8 INSIDE
 * a string -- an I-JSON pathology -> CRYPTO_IJSON_INVALID_STRING -- from bytes
 * that are not valid JSON at all -> the existing CRYPTO_INVALID_JWS_FORMAT code,
 * the same outcome JSON.parse would have produced). It does NOT build a value
 * tree (the existing JSON.parse does that next, on the same decoded bytes). Only
 * the three I-JSON pathologies get the new CRYPTO_IJSON_* codes.
 *
 * No BigInt is used for the numeric range check; range decisions are made by
 * bounded decimal-string comparison.
 *
 * @packageDocumentation
 */

import { CryptoError } from './errors';

/** 2^53 - 1 = Number.MAX_SAFE_INTEGER, as a digit string for magnitude compare. */
const MAX_SAFE_DIGITS = '9007199254740991';

/** Reused fatal UTF-8 decoder for validating raw byte runs inside strings. */
const FATAL_UTF8 = new TextDecoder('utf-8', { fatal: true });

// Byte constants (ASCII).
const TAB = 0x09;
const LF = 0x0a;
const CR = 0x0d;
const SPACE = 0x20;
const QUOTE = 0x22;
const PLUS = 0x2b;
const MINUS = 0x2d;
const DOT = 0x2e;
const ZERO = 0x30;
const NINE = 0x39;
const COLON = 0x3a;
const BACKSLASH = 0x5c;
const LBRACE = 0x7b;
const RBRACE = 0x7d;
const LBRACKET = 0x5b;
const RBRACKET = 0x5d;
const COMMA = 0x2c;

/**
 * Enforce I-JSON (RFC 7493) on the raw UTF-8 bytes of a JSON document.
 * Throws a CryptoError if the bytes violate I-JSON; returns void otherwise.
 *
 * @param bytes - the exact wire bytes of the JSON document (e.g. a base64url-decoded
 *   JWS protected header or payload), validated BEFORE JSON.parse.
 */
export function assertIJson(bytes: Uint8Array): void {
  const scanner = new IJsonScanner(bytes);
  scanner.parseValue();
  scanner.skipWhitespace();
  if (!scanner.atEnd()) {
    throw new CryptoError('CRYPTO_INVALID_JWS_FORMAT', 'I-JSON: trailing data after JSON value');
  }
}

class IJsonScanner {
  private readonly b: Uint8Array;
  private i: number;

  constructor(bytes: Uint8Array) {
    this.b = bytes;
    this.i = 0;
  }

  atEnd(): boolean {
    return this.i >= this.b.length;
  }

  private syntax(msg: string): never {
    throw new CryptoError('CRYPTO_INVALID_JWS_FORMAT', `I-JSON: ${msg}`);
  }

  private invalidString(msg: string): never {
    throw new CryptoError('CRYPTO_IJSON_INVALID_STRING', `I-JSON: ${msg}`);
  }

  skipWhitespace(): void {
    while (this.i < this.b.length) {
      const c = this.b[this.i];
      if (c === SPACE || c === TAB || c === LF || c === CR) this.i++;
      else break;
    }
  }

  parseValue(): void {
    this.skipWhitespace();
    if (this.atEnd()) this.syntax('unexpected end of input');
    const c = this.b[this.i];
    switch (c) {
      case LBRACE:
        this.parseObject();
        return;
      case LBRACKET:
        this.parseArray();
        return;
      case QUOTE:
        this.parseString();
        return;
      case 0x74: // 't'
        this.parseLiteral('true');
        return;
      case 0x66: // 'f'
        this.parseLiteral('false');
        return;
      case 0x6e: // 'n'
        this.parseLiteral('null');
        return;
      default:
        if (c === MINUS || (c >= ZERO && c <= NINE)) {
          this.parseNumber();
          return;
        }
        this.syntax(`unexpected byte 0x${c.toString(16)}`);
    }
  }

  private parseLiteral(lit: string): void {
    for (let k = 0; k < lit.length; k++) {
      if (this.b[this.i + k] !== lit.charCodeAt(k)) {
        this.syntax(`invalid literal, expected ${lit}`);
      }
    }
    this.i += lit.length;
  }

  private parseObject(): void {
    this.i++; // consume '{'
    const names = new Set<string>();
    this.skipWhitespace();
    if (this.b[this.i] === RBRACE) {
      this.i++;
      return;
    }
    for (;;) {
      this.skipWhitespace();
      if (this.b[this.i] !== QUOTE) this.syntax('expected string member name');
      const name = this.parseString();
      if (names.has(name)) {
        throw new CryptoError(
          'CRYPTO_IJSON_DUPLICATE_MEMBER_NAME',
          'I-JSON: duplicate object member name'
        );
      }
      names.add(name);
      this.skipWhitespace();
      if (this.b[this.i] !== COLON) this.syntax("expected ':' after member name");
      this.i++;
      this.parseValue();
      this.skipWhitespace();
      const next = this.b[this.i];
      if (next === COMMA) {
        this.i++;
        continue;
      }
      if (next === RBRACE) {
        this.i++;
        return;
      }
      this.syntax("expected ',' or '}' in object");
    }
  }

  private parseArray(): void {
    this.i++; // consume '['
    this.skipWhitespace();
    if (this.b[this.i] === RBRACKET) {
      this.i++;
      return;
    }
    for (;;) {
      this.parseValue();
      this.skipWhitespace();
      const next = this.b[this.i];
      if (next === COMMA) {
        this.i++;
        continue;
      }
      if (next === RBRACKET) {
        this.i++;
        return;
      }
      this.syntax("expected ',' or ']' in array");
    }
  }

  /**
   * Parse a JSON string from raw bytes: validate escapes + surrogate pairing,
   * validate UTF-8 of raw byte runs, and return the decoded value (used for
   * duplicate-member-name comparison after escape processing per RFC 7493).
   */
  private parseString(): string {
    this.i++; // consume opening '"'
    let decoded = '';
    let runStart = this.i;
    const flushRun = (end: number): void => {
      if (end > runStart) {
        let chunk: string;
        try {
          chunk = FATAL_UTF8.decode(this.b.subarray(runStart, end));
        } catch {
          this.invalidString('invalid UTF-8 byte in string');
        }
        // RFC 7493: reject Unicode noncharacters in directly-encoded UTF-8 too.
        for (const ch of chunk) {
          if (isUnicodeNoncharacter(ch.codePointAt(0)!)) {
            this.invalidString('noncharacter in string');
          }
        }
        decoded += chunk;
      }
    };
    for (;;) {
      if (this.atEnd()) this.invalidString('unterminated string');
      const c = this.b[this.i];
      if (c === QUOTE) {
        flushRun(this.i);
        this.i++;
        return decoded;
      }
      if (c === BACKSLASH) {
        flushRun(this.i);
        this.i++; // consume '\'
        if (this.atEnd()) this.invalidString('unterminated escape sequence');
        const e = this.b[this.i];
        switch (e) {
          case QUOTE:
            decoded += '"';
            this.i++;
            break;
          case BACKSLASH:
            decoded += '\\';
            this.i++;
            break;
          case 0x2f: // '/'
            decoded += '/';
            this.i++;
            break;
          case 0x62: // 'b'
            decoded += '\b';
            this.i++;
            break;
          case 0x66: // 'f'
            decoded += '\f';
            this.i++;
            break;
          case 0x6e: // 'n'
            decoded += '\n';
            this.i++;
            break;
          case 0x72: // 'r'
            decoded += '\r';
            this.i++;
            break;
          case 0x74: // 't'
            decoded += '\t';
            this.i++;
            break;
          case 0x75: {
            // 'u'
            this.i++;
            const hi = this.readHex4();
            if (hi >= 0xd800 && hi <= 0xdbff) {
              if (this.b[this.i] !== BACKSLASH || this.b[this.i + 1] !== 0x75) {
                this.invalidString('lone high surrogate in \\u escape');
              }
              this.i += 2; // consume "\u" of low surrogate
              const lo = this.readHex4();
              if (lo < 0xdc00 || lo > 0xdfff) {
                this.invalidString('high surrogate not followed by a low surrogate');
              }
              const cp = 0x10000 + ((hi - 0xd800) << 10) + (lo - 0xdc00);
              if (isUnicodeNoncharacter(cp)) {
                this.invalidString('noncharacter in string');
              }
              decoded += String.fromCharCode(hi, lo);
            } else if (hi >= 0xdc00 && hi <= 0xdfff) {
              this.invalidString('lone low surrogate in \\u escape');
            } else {
              if (isUnicodeNoncharacter(hi)) {
                this.invalidString('noncharacter in string');
              }
              decoded += String.fromCharCode(hi);
            }
            break;
          }
          default:
            this.invalidString('invalid escape sequence');
        }
        runStart = this.i;
        continue;
      }
      if (c < 0x20) {
        // unescaped control character is not allowed in a JSON string
        this.syntax('unescaped control character in string');
      }
      // Normal byte (printable ASCII, or part of a multibyte UTF-8 sequence
      // validated together by flushRun via the fatal decoder).
      this.i++;
    }
  }

  private readHex4(): number {
    if (this.i + 4 > this.b.length) this.invalidString('truncated \\u escape');
    let v = 0;
    for (let k = 0; k < 4; k++) {
      const ch = this.b[this.i + k];
      let d: number;
      if (ch >= 0x30 && ch <= 0x39)
        d = ch - 0x30; // 0-9
      else if (ch >= 0x41 && ch <= 0x46)
        d = ch - 0x41 + 10; // A-F
      else if (ch >= 0x61 && ch <= 0x66)
        d = ch - 0x61 + 10; // a-f
      else {
        this.invalidString('invalid hex digit in \\u escape');
      }
      v = v * 16 + d;
    }
    this.i += 4;
    return v;
  }

  /** ASCII byte slice -> string (the number literal is pure ASCII). */
  private asciiSlice(start: number, end: number): string {
    let s = '';
    for (let k = start; k < end; k++) s += String.fromCharCode(this.b[k]);
    return s;
  }

  /** Parse a JSON number literal (ASCII) and enforce the I-JSON range predicate. */
  private parseNumber(): void {
    if (this.b[this.i] === MINUS) this.i++;
    // integer part
    const intStart = this.i;
    if (this.b[this.i] === ZERO) {
      this.i++;
    } else if (this.b[this.i] >= 0x31 && this.b[this.i] <= NINE) {
      while (this.b[this.i] >= ZERO && this.b[this.i] <= NINE) this.i++;
    } else {
      this.syntax('invalid number: missing integer digits');
    }
    const intDigits = this.asciiSlice(intStart, this.i);
    // fraction part
    let fracDigits = '';
    if (this.b[this.i] === DOT) {
      this.i++;
      const fracStart = this.i;
      if (!(this.b[this.i] >= ZERO && this.b[this.i] <= NINE)) {
        this.syntax('invalid number: missing fraction digits');
      }
      while (this.b[this.i] >= ZERO && this.b[this.i] <= NINE) this.i++;
      fracDigits = this.asciiSlice(fracStart, this.i);
    }
    // exponent part (kept as digits + sign; never parsed into a JS number, so a
    // huge exponent cannot overflow or trigger a large allocation)
    let expDigits = '';
    let expNegative = false;
    if (this.b[this.i] === 0x65 || this.b[this.i] === 0x45) {
      // 'e' | 'E'
      this.i++;
      if (this.b[this.i] === PLUS) this.i++;
      else if (this.b[this.i] === MINUS) {
        expNegative = true;
        this.i++;
      }
      const expStart = this.i;
      if (!(this.b[this.i] >= ZERO && this.b[this.i] <= NINE)) {
        this.syntax('invalid number: missing exponent digits');
      }
      while (this.b[this.i] >= ZERO && this.b[this.i] <= NINE) this.i++;
      expDigits = this.asciiSlice(expStart, this.i);
    }

    if (numberMagnitudeExceedsSafe(intDigits, fracDigits, expDigits, expNegative)) {
      throw new CryptoError(
        'CRYPTO_IJSON_NUMBER_OUT_OF_RANGE',
        'I-JSON: number is non-finite or exceeds the safe numeric range; encode large values as strings'
      );
    }
  }
}

/**
 * I-JSON number magnitude predicate. Returns true if the number must be REJECTED
 * because its absolute decimal magnitude is greater than 9007199254740991
 * (= 2^53 - 1). Exact for every JSON number form (integer, fraction, exponent):
 * the comparison is performed on the decimal digit string against the bound, so
 * binary64 rounding can never accept a value whose exact magnitude exceeds the
 * bound (e.g. 9007199254740991.1). No BigInt is used; range decisions are made by
 * bounded decimal-string comparison. The exponent is consumed as digits + sign
 * and clamped, so an arbitrarily large exponent cannot overflow or allocate.
 *
 * value = M * 10^decimalExp, where M is the significand digits (leading zeros
 * stripped) and decimalExp = exp - len(fracDigits). intDigitCount is the number
 * of integer digits of the magnitude.
 */
function numberMagnitudeExceedsSafe(
  intDigits: string,
  fracDigits: string,
  expDigits: string,
  expNegative: boolean
): boolean {
  const sig = stripLeadingZeros(intDigits + fracDigits);
  if (sig === '0') return false; // value is exactly 0
  const L = sig.length;

  const decimalExp = clampedExp(expDigits, expNegative) - fracDigits.length;
  const intDigitCount = L + decimalExp; // integer digits of the magnitude

  const BOUND = MAX_SAFE_DIGITS; // 16 digits
  if (intDigitCount > BOUND.length) return true; // >= 10^16 > bound
  if (intDigitCount < BOUND.length) return false; // < 10^15 < bound

  // intDigitCount === 16: compare the 16-digit integer part exactly.
  if (decimalExp >= 0) {
    // integer value: M right-padded with decimalExp zeros (total length 16).
    const intPart = sig + '0'.repeat(decimalExp);
    return intPart > BOUND;
  }
  // decimalExp < 0: integer part is the first 16 digits; remainder is fractional.
  const intPart = sig.slice(0, BOUND.length);
  if (intPart > BOUND) return true;
  if (intPart < BOUND) return false;
  // integer part == bound: any nonzero fractional digit pushes magnitude over.
  return /[1-9]/.test(sig.slice(BOUND.length));
}

/**
 * Parse an exponent digit string + sign into a JS integer, clamped so a very
 * large exponent cannot overflow. The clamp magnitude (1e9) far exceeds any
 * possible significand length for size-bounded inputs, so a clamped exponent
 * always drives intDigitCount past the decision boundary in the correct direction.
 */
function clampedExp(expDigits: string, expNegative: boolean): number {
  if (expDigits.length === 0) return 0;
  // Strip leading zeros BEFORE the length-based clamp so a zero-padded exponent
  // (e.g. 1e0000000003 == 1e3) is not mistaken for a huge exponent.
  const normalized = stripLeadingZeros(expDigits);
  if (normalized === '0') return 0;
  const mag = normalized.length > 9 ? 1_000_000_000 : Number(normalized);
  return expNegative ? -mag : mag;
}

function stripLeadingZeros(digits: string): string {
  let j = 0;
  while (j < digits.length - 1 && digits[j] === '0') j++;
  return digits.slice(j);
}

/**
 * RFC 7493: a JSON string (and member name) MUST NOT contain Unicode
 * noncharacters, in directly-encoded or escaped form. The 66 noncharacters are
 * U+FDD0..U+FDEF and the last two code points of every plane (U+xFFFE, U+xFFFF).
 */
function isUnicodeNoncharacter(cp: number): boolean {
  return (
    (cp >= 0xfdd0 && cp <= 0xfdef) ||
    (cp >= 0xfffe && cp <= 0x10ffff && ((cp & 0xffff) === 0xfffe || (cp & 0xffff) === 0xffff))
  );
}

File: sdks/go/ijson.go
package peac

// I-JSON (RFC 7493) raw-bytes gate.
//
// Runs on the UTF-8 bytes of a JSON document BEFORE json.Unmarshal, because
// platform JSON parsers can collapse duplicate object member names, round or
// overflow numeric values, or substitute invalid string data before higher-level
// validation sees the original bytes. The gate also classifies string-level
// syntax failures consistently at the raw boundary. It rejects, per I-JSON
// (RFC 7493):
//
//   - duplicate object member names  -> json.Unmarshal keeps the last; Token()
//     emits both but neither rejects   => E_IJSON_DUPLICATE_MEMBER_NAME
//   - numbers non-finite OR whose absolute magnitude exceeds 2^53-1 (incl. large
//     finite floats) -> rounded/overflowed => E_IJSON_NUMBER_OUT_OF_RANGE
//   - invalid string content: lone surrogates or noncharacters (directly encoded
//     or in \u escapes), invalid escape sequences, or invalid UTF-8 bytes
//     inside a string. NOTE: Go's
//     json.Decoder.Token() SILENTLY replaces lone surrogates and invalid UTF-8
//     with U+FFFD (never errors), so it cannot be relied on for I-JSON string
//     validity. => E_IJSON_INVALID_STRING
//
// This is a strict single-pass byte-level JSON parser (a direct port of the
// TypeScript packages/crypto/src/ijson.ts scanner, for maximum cross-language
// parity). It distinguishes invalid UTF-8 INSIDE a string (an I-JSON pathology
// -> E_IJSON_INVALID_STRING) from bytes that are not valid JSON at all
// (-> E_INVALID_FORMAT, the same outcome json.Unmarshal would have produced).
// It does NOT build a value tree (the existing json.Unmarshal does that next).
//
// The gate is internal (unexported); no big.Int is used for the numeric range
// check; range decisions are made by bounded decimal-string comparison.

import (
	"fmt"
	"strings"
	"unicode/utf8"
)

// maxSafeDigits is 2^53 - 1 (= JS Number.MAX_SAFE_INTEGER), as a digit string.
const maxSafeDigits = "9007199254740991"

// ijsonError is a typed error from the I-JSON gate. Code is a canonical PEAC
// error code: one of E_IJSON_DUPLICATE_MEMBER_NAME, E_IJSON_NUMBER_OUT_OF_RANGE,
// E_IJSON_INVALID_STRING (the three I-JSON pathologies), or E_INVALID_FORMAT
// (generic JSON syntax error, same as json.Unmarshal would produce).
type ijsonError struct {
	Code string
	Msg  string
}

func (e *ijsonError) Error() string { return e.Msg }

// assertIJSON enforces I-JSON (RFC 7493) on the raw UTF-8 bytes of a JSON
// document. It returns an *ijsonError if the bytes violate I-JSON, nil otherwise.
func assertIJSON(data []byte) error {
	s := &ijsonScanner{b: data}
	if err := s.parseValue(); err != nil {
		return err
	}
	s.skipWhitespace()
	if !s.atEnd() {
		return s.syntax("trailing data after JSON value")
	}
	return nil
}

type ijsonScanner struct {
	b []byte
	i int
}

func (s *ijsonScanner) atEnd() bool { return s.i >= len(s.b) }

func (s *ijsonScanner) syntax(msg string) *ijsonError {
	return &ijsonError{Code: "E_INVALID_FORMAT", Msg: "I-JSON: " + msg}
}

func (s *ijsonScanner) invalidString(msg string) *ijsonError {
	return &ijsonError{Code: "E_IJSON_INVALID_STRING", Msg: "I-JSON: " + msg}
}

func (s *ijsonScanner) skipWhitespace() {
	for s.i < len(s.b) {
		c := s.b[s.i]
		if c == 0x20 || c == 0x09 || c == 0x0a || c == 0x0d {
			s.i++
		} else {
			break
		}
	}
}

func (s *ijsonScanner) parseValue() error {
	s.skipWhitespace()
	if s.atEnd() {
		return s.syntax("unexpected end of input")
	}
	c := s.b[s.i]
	switch {
	case c == '{':
		return s.parseObject()
	case c == '[':
		return s.parseArray()
	case c == '"':
		_, err := s.parseString()
		return err
	case c == 't':
		return s.parseLiteral("true")
	case c == 'f':
		return s.parseLiteral("false")
	case c == 'n':
		return s.parseLiteral("null")
	case c == '-' || (c >= '0' && c <= '9'):
		return s.parseNumber()
	default:
		return s.syntax(fmt.Sprintf("unexpected byte 0x%x", c))
	}
}

func (s *ijsonScanner) parseLiteral(lit string) error {
	for k := 0; k < len(lit); k++ {
		if s.i+k >= len(s.b) || s.b[s.i+k] != lit[k] {
			return s.syntax("invalid literal, expected " + lit)
		}
	}
	s.i += len(lit)
	return nil
}

func (s *ijsonScanner) parseObject() error {
	s.i++ // consume '{'
	names := make(map[string]struct{})
	s.skipWhitespace()
	if s.i < len(s.b) && s.b[s.i] == '}' {
		s.i++
		return nil
	}
	for {
		s.skipWhitespace()
		if s.i >= len(s.b) || s.b[s.i] != '"' {
			return s.syntax("expected string member name")
		}
		name, err := s.parseString()
		if err != nil {
			return err
		}
		if _, dup := names[name]; dup {
			return &ijsonError{
				Code: "E_IJSON_DUPLICATE_MEMBER_NAME",
				Msg:  "I-JSON: duplicate object member name",
			}
		}
		names[name] = struct{}{}
		s.skipWhitespace()
		if s.i >= len(s.b) || s.b[s.i] != ':' {
			return s.syntax("expected ':' after member name")
		}
		s.i++
		if err := s.parseValue(); err != nil {
			return err
		}
		s.skipWhitespace()
		if s.atEnd() {
			return s.syntax("expected ',' or '}' in object")
		}
		switch s.b[s.i] {
		case ',':
			s.i++
		case '}':
			s.i++
			return nil
		default:
			return s.syntax("expected ',' or '}' in object")
		}
	}
}

func (s *ijsonScanner) parseArray() error {
	s.i++ // consume '['
	s.skipWhitespace()
	if s.i < len(s.b) && s.b[s.i] == ']' {
		s.i++
		return nil
	}
	for {
		if err := s.parseValue(); err != nil {
			return err
		}
		s.skipWhitespace()
		if s.atEnd() {
			return s.syntax("expected ',' or ']' in array")
		}
		switch s.b[s.i] {
		case ',':
			s.i++
		case ']':
			s.i++
			return nil
		default:
			return s.syntax("expected ',' or ']' in array")
		}
	}
}

// parseString parses a JSON string from raw bytes: validates escapes + surrogate
// pairing + UTF-8 of raw runs, and returns the decoded value (used for
// duplicate-member-name comparison after escape processing per RFC 7493).
func (s *ijsonScanner) parseString() (string, error) {
	s.i++ // consume opening '"'
	var out strings.Builder
	runStart := s.i
	flushRun := func(end int) error {
		if end > runStart {
			run := s.b[runStart:end]
			if !utf8.Valid(run) {
				return s.invalidString("invalid UTF-8 byte in string")
			}
			// RFC 7493: reject Unicode noncharacters in directly-encoded UTF-8 too.
			for _, r := range string(run) {
				if isUnicodeNoncharacter(r) {
					return s.invalidString("noncharacter in string")
				}
			}
			out.Write(run)
		}
		return nil
	}
	for {
		if s.atEnd() {
			return "", s.invalidString("unterminated string")
		}
		c := s.b[s.i]
		if c == '"' {
			if err := flushRun(s.i); err != nil {
				return "", err
			}
			s.i++
			return out.String(), nil
		}
		if c == '\\' {
			if err := flushRun(s.i); err != nil {
				return "", err
			}
			s.i++ // consume '\'
			if s.atEnd() {
				return "", s.invalidString("unterminated escape sequence")
			}
			e := s.b[s.i]
			switch e {
			case '"':
				out.WriteByte('"')
				s.i++
			case '\\':
				out.WriteByte('\\')
				s.i++
			case '/':
				out.WriteByte('/')
				s.i++
			case 'b':
				out.WriteByte('\b')
				s.i++
			case 'f':
				out.WriteByte('\f')
				s.i++
			case 'n':
				out.WriteByte('\n')
				s.i++
			case 'r':
				out.WriteByte('\r')
				s.i++
			case 't':
				out.WriteByte('\t')
				s.i++
			case 'u':
				s.i++
				hi, err := s.readHex4()
				if err != nil {
					return "", err
				}
				if hi >= 0xd800 && hi <= 0xdbff {
					if s.i+1 >= len(s.b) || s.b[s.i] != '\\' || s.b[s.i+1] != 'u' {
						return "", s.invalidString("lone high surrogate in \\u escape")
					}
					s.i += 2 // consume "\u" of low surrogate
					lo, err := s.readHex4()
					if err != nil {
						return "", err
					}
					if lo < 0xdc00 || lo > 0xdfff {
						return "", s.invalidString("high surrogate not followed by a low surrogate")
					}
					r := 0x10000 + (rune(hi-0xd800) << 10) + rune(lo-0xdc00)
					if isUnicodeNoncharacter(r) {
						return "", s.invalidString("noncharacter in string")
					}
					out.WriteRune(r)
				} else if hi >= 0xdc00 && hi <= 0xdfff {
					return "", s.invalidString("lone low surrogate in \\u escape")
				} else {
					if isUnicodeNoncharacter(rune(hi)) {
						return "", s.invalidString("noncharacter in string")
					}
					out.WriteRune(rune(hi))
				}
			default:
				return "", s.invalidString("invalid escape sequence")
			}
			runStart = s.i
			continue
		}
		if c < 0x20 {
			return "", s.syntax("unescaped control character in string")
		}
		// Normal byte (printable ASCII, or part of a multibyte UTF-8 sequence
		// validated together by flushRun via utf8.Valid).
		s.i++
	}
}

func (s *ijsonScanner) readHex4() (int, error) {
	if s.i+4 > len(s.b) {
		return 0, s.invalidString("truncated \\u escape")
	}
	v := 0
	for k := 0; k < 4; k++ {
		ch := s.b[s.i+k]
		var d int
		switch {
		case ch >= '0' && ch <= '9':
			d = int(ch - '0')
		case ch >= 'A' && ch <= 'F':
			d = int(ch-'A') + 10
		case ch >= 'a' && ch <= 'f':
			d = int(ch-'a') + 10
		default:
			return 0, s.invalidString("invalid hex digit in \\u escape")
		}
		v = v*16 + d
	}
	s.i += 4
	return v, nil
}

func (s *ijsonScanner) asciiSlice(start, end int) string {
	return string(s.b[start:end])
}

// parseNumber parses a JSON number literal (ASCII) and enforces the I-JSON range predicate.
func (s *ijsonScanner) parseNumber() error {
	if s.b[s.i] == '-' {
		s.i++
	}
	// integer part
	intStart := s.i
	if s.i < len(s.b) && s.b[s.i] == '0' {
		s.i++
	} else if s.i < len(s.b) && s.b[s.i] >= '1' && s.b[s.i] <= '9' {
		for s.i < len(s.b) && s.b[s.i] >= '0' && s.b[s.i] <= '9' {
			s.i++
		}
	} else {
		return s.syntax("invalid number: missing integer digits")
	}
	intDigits := s.asciiSlice(intStart, s.i)
	// fraction part
	fracDigits := ""
	if s.i < len(s.b) && s.b[s.i] == '.' {
		s.i++
		fracStart := s.i
		if s.i >= len(s.b) || s.b[s.i] < '0' || s.b[s.i] > '9' {
			return s.syntax("invalid number: missing fraction digits")
		}
		for s.i < len(s.b) && s.b[s.i] >= '0' && s.b[s.i] <= '9' {
			s.i++
		}
		fracDigits = s.asciiSlice(fracStart, s.i)
	}
	// exponent part (kept as digits + sign; never parsed into an int that could
	// overflow, and never used to allocate)
	expDigits := ""
	expNegative := false
	if s.i < len(s.b) && (s.b[s.i] == 'e' || s.b[s.i] == 'E') {
		s.i++
		if s.i < len(s.b) && s.b[s.i] == '+' {
			s.i++
		} else if s.i < len(s.b) && s.b[s.i] == '-' {
			expNegative = true
			s.i++
		}
		expStart := s.i
		if s.i >= len(s.b) || s.b[s.i] < '0' || s.b[s.i] > '9' {
			return s.syntax("invalid number: missing exponent digits")
		}
		for s.i < len(s.b) && s.b[s.i] >= '0' && s.b[s.i] <= '9' {
			s.i++
		}
		expDigits = s.asciiSlice(expStart, s.i)
	}

	if numberMagnitudeExceedsSafe(intDigits, fracDigits, expDigits, expNegative) {
		return &ijsonError{
			Code: "E_IJSON_NUMBER_OUT_OF_RANGE",
			Msg:  "I-JSON: number is non-finite or exceeds the safe numeric range; encode large values as strings",
		}
	}
	return nil
}

// numberMagnitudeExceedsSafe reports whether a number must be REJECTED because
// its absolute decimal magnitude is greater than 2^53-1. Exact for every JSON
// number form (integer, fraction, exponent): the comparison is performed on the
// decimal digit string against the bound, so binary64 rounding can never accept a
// value whose exact magnitude exceeds the bound (e.g. 9007199254740991.1). No
// big.Int is used. The exponent is consumed as digits + sign and clamped, so an
// arbitrarily large exponent cannot overflow or allocate.
//
// value = M * 10^decimalExp, where M is the significand digits (leading zeros
// stripped) and decimalExp = exp - len(fracDigits). intDigitCount is the number
// of integer digits of the magnitude.
func numberMagnitudeExceedsSafe(intDigits, fracDigits, expDigits string, expNegative bool) bool {
	sig := stripLeadingZeros(intDigits + fracDigits)
	if sig == "0" {
		return false // value is exactly 0
	}
	L := len(sig)

	decimalExp := clampedExp(expDigits, expNegative) - len(fracDigits)
	intDigitCount := L + decimalExp // integer digits of the magnitude

	bound := maxSafeDigits // 16 digits
	if intDigitCount > len(bound) {
		return true // >= 10^16 > bound
	}
	if intDigitCount < len(bound) {
		return false // < 10^15 < bound
	}

	// intDigitCount == 16: compare the 16-digit integer part exactly.
	if decimalExp >= 0 {
		intPart := sig + strings.Repeat("0", decimalExp)
		return intPart > bound
	}
	// decimalExp < 0: integer part is the first 16 digits; remainder is fractional.
	intPart := sig[:len(bound)]
	if intPart > bound {
		return true
	}
	if intPart < bound {
		return false
	}
	// integer part == bound: any nonzero fractional digit pushes magnitude over.
	for _, ch := range sig[len(bound):] {
		if ch != '0' {
			return true
		}
	}
	return false
}

// clampedExp parses an exponent digit string + sign into an int, clamped so a
// very large exponent cannot overflow. The clamp magnitude (1e9) far exceeds any
// possible significand length for size-bounded inputs, so a clamped exponent
// always drives intDigitCount past the decision boundary in the correct direction.
func clampedExp(expDigits string, expNegative bool) int {
	if len(expDigits) == 0 {
		return 0
	}
	// Strip leading zeros BEFORE the length-based clamp so a zero-padded exponent
	// (e.g. 1e0000000003 == 1e3) is not mistaken for a huge exponent.
	normalized := stripLeadingZeros(expDigits)
	if normalized == "0" {
		return 0
	}
	mag := 1_000_000_000
	if len(normalized) <= 9 {
		mag = 0
		for i := 0; i < len(normalized); i++ {
			mag = mag*10 + int(normalized[i]-'0')
		}
	}
	if expNegative {
		return -mag
	}
	return mag
}

func stripLeadingZeros(digits string) string {
	j := 0
	for j < len(digits)-1 && digits[j] == '0' {
		j++
	}
	return digits[j:]
}

// isUnicodeNoncharacter reports whether a code point is a Unicode noncharacter.
// RFC 7493: a JSON string (and member name) MUST NOT contain noncharacters, in
// directly-encoded or escaped form. The 66 noncharacters are U+FDD0..U+FDEF and
// the last two code points of every plane (U+xFFFE, U+xFFFF).
func isUnicodeNoncharacter(r rune) bool {
	return (r >= 0xFDD0 && r <= 0xFDEF) ||
		(r >= 0xFFFE && r <= 0x10FFFF && ((r&0xFFFF) == 0xFFFE || (r&0xFFFF) == 0xFFFF))
}
